base/shell/explorer/startmnucust.cpp

```
@@ -58,10 +58,53 @@ static BOOL RecentHasShortcut(HWND hwnd)
     return TRUE;
 }
 
-static VOID OnClearRecentItems(HWND hwnd)
+static const PCWSTR g_MruKeys[] = 
+{
+    L"Software\\Microsoft\\Internet Explorer\\TypedURLs",
+    L"Explorer\\RunMRU",
+    L"Explorer\\Comdlg32\\OpenSaveMRU",
+    L"Explorer\\Comdlg32\\LastVisitedMRU",
+};
+
+static BOOL HandleMruData(BOOL Delete)
+{
+    for (UINT i = 0; i < _countof(g_MruKeys); ++i)
+    {
+        WCHAR szKey[200];
+        PCWSTR pszKey = g_MruKeys[i];
+        if (*pszKey != 'S') // Keys not starting with S[oftware] are assumed to be relative to "SMWCV"
+        {
+            wsprintfW(szKey, L"%s\\%s", L"Software\\Microsoft\\Windows\\CurrentVersion", pszKey);
+            pszKey = szKey;
+        }
+
+        HKEY hKey;
+        if (Delete)
+        {
+            SHDeleteKeyW(HKEY_CURRENT_USER, pszKey);
+        }
+        else if (RegOpenKeyExW(HKEY_CURRENT_USER, pszKey, 0, KEY_WRITE, &hKey) == ERROR_SUCCESS)
+        {
+            RegCloseKey(hKey);
+            return TRUE;
+        }
+    }
+    return FALSE;
```

Weird to return FALSE "by default" (case of Delete == TRUE for example).

---

We don't care about the return value of the delete operation. We do however care about the keys existing or not.

---

> We don't care about the return value of the delete operation

yet.

---

Never?

How can it fail?

- The key is deleted by another process. We would just ignore this error anyway.
- Another registry error? Somebody changed the ACL so we can't delete our own keys in HKCU? How would we explain this to the user?

Worst case, if we can't delete a key, the Clear button just stays enabled in the UI.